### PR TITLE
Create dedicated build for training api

### PR DESCRIPTION
**Pull request by askhade**
### Description
Enable creating dedicated build for on device training. With this PR we can build a lean binary for on device training using flag --enable_training_apis. This binary includes only the essentials like training ops, optimizers etc and NOT features like Aten fallback, strided tensors, gradient builders etc . This binary also removes all the deprecated components like training::TrainingSession and OrtTrainer etc

### Motivation and Context
This enables our partners to create a lean binary for on device training.

### CODE REVIEW

File: onnxruntime/core/framework/session_state.cc [hunk 1]
```diff
@@ -958,7 +958,7 @@ const NodeIndexInfo& SessionState::GetNodeIndexInfo() const {
   return *node_index_info_;
 }
 
-#ifdef ENABLE_TRAINING
+#ifdef ENABLE_TRAINING_CORE
 void SessionState::UpdateToBeExecutedRange(gsl::span<int const> fetch_mlvalue_idxs) {
```
Comment (souptc) [inline onnxruntime/core/framework/session_state.cc:962]: this is also for partial execution.

File: onnxruntime/core/framework/session_state.h [hunk 2]
```diff
@@ -39,7 +39,7 @@
 #endif
 
 #include "core/framework/stream_handles.h"
-#ifdef ENABLE_TRAINING
+#ifdef ENABLE_TRAINING_CORE
 #include "core/framework/program_region.h"
```
Comment (souptc) [inline onnxruntime/core/framework/session_state.h:43]: this is also for partial execution.

File: onnxruntime/core/framework/session_state.h [hunk 3]
```diff
@@ -515,7 +515,7 @@ class SessionState {
   // prepacked_weights_container_ can be nullptr if no caching is required for prepacked weights
   PrepackedWeightsContainer* const prepacked_weights_container_{};
 
-#ifdef ENABLE_TRAINING
+#ifdef ENABLE_TRAINING_CORE
 #ifndef DISABLE_ABSEIL
   InlinedHashMap<InlinedVector<int>, InlinedHashSet<NodeIndex>> to_be_executed_nodes_;
```
Comment (souptc) [inline onnxruntime/core/framework/session_state.h:520]: this is for orttrainer. maybe we can put a comment here, to remove it when remove orttrainer.

File: onnxruntime/core/optimizer/compute_optimizer/compute_optimizer.cc [hunk 4]
```diff
@@ -1,7 +1,7 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
-#ifdef ENABLE_TRAINING
+#ifdef ENABLE_TRAINING_CORE
 #include <onnx/defs/attr_proto_util.h>
```
Comment (souptc) [inline onnxruntime/core/optimizer/compute_optimizer/compute_optimizer.cc:5]: if the whole file is only for training build, could we just put it under the training folder, and only include it in training build in the cmake, instead of have the ifdef in the code?

File: onnxruntime/core/optimizer/graph_transformer_utils.cc [hunk 5]
```diff
@@ -67,12 +67,14 @@
...
 #include "core/optimizer/identical_children_consolidation.h"
-#ifdef ENABLE_TRAINING
+#ifdef ENABLE_TRAINING_CORE
 #include "orttraining/core/optimizer/bitmask_dropout_replacement.h"
 #include "orttraining/core/optimizer/bias_softmax_dropout_fusion.h"
-#include "orttraining/core/optimizer/memory_optimizer.h"
 #include "orttraining/core/optimizer/sce_loss_grad_bias_fusion.h"
 #endif
+#ifdef ENABLE_TRAINING
+#include "orttraining/core/optimizer/memory_optimizer.h"
```
Comment (souptc) [inline onnxruntime/core/optimizer/graph_transformer_utils.cc:76]: you don't need the recompute feature in training_api build?

File: onnxruntime/core/session/environment.cc [hunk 6]
```diff
@@ -34,13 +34,14 @@
 #if defined(ENABLE_TRAINING_OPS)
 #include "orttraining/core/graph/training_op_defs.h"
 #endif
+#ifdef ENABLE_TRAINING_CORE
+#include "orttraining/core/graph/loss_function_registry.h"
```
Comment (souptc) [inline onnxruntime/core/session/environment.cc:38]: do you need these for ORTModule build? I thought it is only useful for training c++ api, but ENABLE_TRAINING_CORE seems to be shared by both side.

File: cmake/CMakeLists.txt [hunk 7]
```diff
@@ -1301,8 +1324,9 @@ if (onnxruntime_ENABLE_ROCM_PROFILING)
 endif()
 
 if (onnxruntime_ENABLE_TRAINING)
-  add_compile_definitions(ENABLE_TRAINING)
+  add_compile_definitions(ENABLE_TRAINING_CORE)
   add_compile_definitions(ENABLE_STRIDED_TENSORS)
+  add_compile_definitions(ENABLE_TRAINING)
```
[the remark this answers is not in the document]
Comment (askhade) [reply cmake/CMakeLists.txt:1329]: We do need it right now... This flag is used to include code related to training full build and includes features like partial executor etc along with some deprecated features like ORTTRainer, training::TrainingSession etc. More refactoring is necessary to make this only a user facing macro.

File: cmake/external/onnxruntime_external_deps.cmake [hunk 8]
```diff
@@ -429,7 +429,7 @@ if(onnxruntime_ENABLE_ATEN)
   FetchContent_Populate(dlpack)
 endif()
 
-if(onnxruntime_ENABLE_TRAINING)
+if(onnxruntime_ENABLE_TRAINING OR (onnxruntime_ENABLE_TRAINING_APIS AND onnxruntime_BUILD_UNIT_TESTS))
```
[the remark this answers is not in the document]
Comment (askhade) [reply cmake/external/onnxruntime_external_deps.cmake:432]: Yes. It is being used in onnxruntime_training_runner executable as well as onnxruntime_training_mnist and onnxruntime_training_gpt2... All this is deprecated code and we can simply remove onnxruntime_ENABLE_TRAINING from there once that code is removed. I will add this as a comment in the cmake.

File: onnxruntime/core/framework/allocation_planner.cc [hunk 9]
```diff
@@ -1476,7 +1476,7 @@ class PlannerImpl {
     return Status::OK();
   }
 
-#ifdef ENABLE_TRAINING
+#ifdef ENABLE_TRAINING_CORE
   Status CalculateProgramCounter() {
```
[the remark this answers is not in the document]
Comment (askhade) [reply onnxruntime/core/framework/allocation_planner.cc:1480]: I am moving this back to ENABLE_TRAINING... let's discuss offline how to reduce the complexity and create a new PR to handle it.

File: onnxruntime/core/framework/sequential_executor.cc [hunk 10]
```diff
@@ -441,10 +441,12 @@ onnxruntime::Status ExecuteKernel(StreamExecutionContext& ctx,
...
     KernelScope kernel_scope(session_scope, kernel_ctx, *p_kernel);
     ORT_TRY {
-#ifdef ENABLE_TRAINING
+#ifdef ENABLE_TRAINING_CORE
       if (p_kernel->KernelDef().AllocateInputsContiguously()) {
         ORT_RETURN_IF_ERROR(utils::VerifyInputTensorsAllocatedContiguously(&kernel_ctx));
       }
+#endif
+#ifdef ENABLE_TRAINING
       // Cache lookup. Currently we only cache single-output nodes,
```
[the remark this answers is not in the document]
Comment (askhade) [reply onnxruntime/core/framework/sequential_executor.cc:450]: Agree... will create a new PR for this.

File: onnxruntime/core/optimizer/graph_transformer_utils.cc (hunk 5 above)
Comment (askhade) [reply onnxruntime/core/optimizer/graph_transformer_utils.cc:76]: No, current target models which will use training api will be very simple... this optimizer is required for advanced scenarios when we want to increase the batch size by saving some memory and recomputing the activations again when it is time for gradient computation.

File: onnxruntime/core/session/environment.cc (hunk 6 above)
Comment (askhade) [reply onnxruntime/core/session/environment.cc:38]: I think they are needed (will check). However, we dont do any dedicated build for ortmodule, we simply do a full training build, so will need these.

File: cmake/onnxruntime_graph.cmake [hunk 11]
```diff
@@ -72,6 +72,27 @@ if (onnxruntime_ENABLE_TRAINING_OPS AND NOT onnxruntime_ENABLE_TRAINING)
       "${ORTTRAINING_SOURCE_DIR}/core/graph/training_op_defs.h"
       )
 endif()
+
+if (onnxruntime_ENABLE_TRAINING_APIS AND NOT onnxruntime_ENABLE_TRAINING)
```
[the remark this answers is not in the document]
Comment (pengwa) [reply cmake/onnxruntime_graph.cmake:76]: I believe onnxruntime_ENABLE_TRAINING_APIS  is for on-device training, so it does not want to include those ORT trainer C++ codes. But yes, onnxruntime_ENABLE_TRAINING_APIS from its name is a subset of ENABLE_TRAINING, I recalled it is called TRAINING_ON_DEVICE previously.

File: cmake/onnxruntime_providers.cmake [hunk 12]
```diff
@@ -471,7 +471,9 @@ if (onnxruntime_USE_CUDA)
   onnxruntime_add_include_to_target(onnxruntime_providers_cuda onnxruntime_common onnxruntime_framework onnx onnx_proto ${PROTOBUF_LIB} flatbuffers::flatbuffers)
   if (onnxruntime_ENABLE_TRAINING_OPS)
     onnxruntime_add_include_to_target(onnxruntime_providers_cuda onnxruntime_training)
-    target_link_libraries(onnxruntime_providers_cuda PRIVATE onnxruntime_training)
+    if (onnxruntime_ENABLE_TRAINING)
```
Comment (pengwa) [inline cmake/onnxruntime_providers.cmake:474]: if "if (onnxruntime_ENABLE_TRAINING_OPS)" is true, then "if (onnxruntime_ENABLE_TRAINING)" is ture, right?

File: cmake/onnxruntime_unittests.cmake [hunk 13]
```diff
@@ -1283,6 +1291,55 @@ if (NOT onnxruntime_ENABLE_TRAINING_TORCH_INTEROP)
...
+      ${ORTTRAINING_ROOT}
+      ${eigen_INCLUDE_DIRS}
+      ${CXXOPTS}
+      ${extra_includes}
+      ${onnxruntime_graph_header}
+      ${onnxruntime_exec_src_dir}
+    )
+
+    set(ONNXRUNTIME_TEST_LIBS
+    onnxruntime_session
```
Comment (pengwa) [inline cmake/onnxruntime_unittests.cmake:1323]: format is off

File: onnxruntime/core/framework/session_state.h [hunk 14]
```diff
@@ -516,6 +516,7 @@ class SessionState {
   PrepackedWeightsContainer* const prepacked_weights_container_{};
 
 #ifdef ENABLE_TRAINING
+// Needed for ORTTRainer. Should be removed along with ORTTrainer code
```
Comment (pengwa) [inline onnxruntime/core/framework/session_state.h:519]: nit: ORTTrainer

File: onnxruntime/core/optimizer/compute_optimizer/compute_optimizer.cc (hunk 4 above)
Comment (pengwa) [reply onnxruntime/core/optimizer/compute_optimizer/compute_optimizer.cc:5]: this is added by me, I put it here for easily enabling it for inferencing later.

File: onnxruntime/core/session/environment.cc (hunk 6 above)
Comment (pengwa) [reply onnxruntime/core/session/environment.cc:38]: I believe  "orttraining/core/graph/loss_function_registry.h" is not needed for training_api (on-device training)

File: onnxruntime/core/session/environment.cc [hunk 15]
```diff
@@ -264,13 +265,16 @@ Status Environment::Initialize(std::unique_ptr<logging::LoggingManager> logging_
       // preserve this order until <training schemas>: this depends on operatorsetschema registration.
       training::RegisterTrainingOpSchemas();
 #endif
+#ifdef ENABLE_TRAINING_CORE
+      training::LossFunctionRegistry::GetInstance().RegisterNonOperatorLossFunctions();
```
Comment (pengwa) [inline onnxruntime/core/session/environment.cc:269]: those loss functions are not useful for ondevice training.

File: onnxruntime/test/contrib_ops/attention_op_test.cc [hunk 16]
```diff
@@ -3650,7 +3650,7 @@ TEST(AttentionTest, DISABLED_Attention_Mask1D_Fp16_B2_FusedNoPadding) {
   }
 }
 
-#ifndef ENABLE_TRAINING  // Prepacking is enabled only on non-training builds
+#ifndef ENABLE_TRAINING_CORE  // Prepacking is enabled only on non-training builds
```
Comment (pengwa) [inline onnxruntime/test/contrib_ops/attention_op_test.cc:3653]: I think we need make it clear the relationship between those training macros somewhere. The inferencing guys at least need to know what macro should be used to explicitly turning off some code snippet.

File: cmake/onnxruntime_providers.cmake (hunk 12 above)
Comment (askhade) [reply cmake/onnxruntime_providers.cmake:474]: not always... 
1. On Device Training build uses enable_training_apis flag and in this case enable_training will not be true
2. There is 1 more scenario where training ops are included in the inference build... I am not 100% sure how it is used but this enable_trianing_ops macro was first added for this scenario.

File: cmake/onnxruntime_providers.cmake [hunk 17]
```diff
@@ -471,7 +471,9 @@ if (onnxruntime_USE_CUDA)
...
+      target_link_libraries(onnxruntime_providers_cuda PRIVATE onnxruntime_training)
```
[the remark this answers is not in the document]
Comment (askhade) [reply cmake/onnxruntime_providers.cmake:475]: gradient graph builder and a lot of deprecated code.

[the remark this answers is not in the document]
Comment (souptc) [reply cmake/onnxruntime_providers.cmake:475]: why cuda EP need to linked with gradient builder / training agent? I thought in training build, the only impact to cuda ep is to include the additional training kernels.

File: onnxruntime/core/session/environment.cc [hunk 18]
```diff
@@ -264,13 +265,16 @@ Status Environment::Initialize(std::unique_ptr<logging::LoggingManager> logging_
       // preserve this order until <training schemas>: this depends on operatorsetschema registration.
       training::RegisterTrainingOpSchemas();
 #endif
+#ifdef ENABLE_TRAINING_CORE
+      // <training schemas>
+      // This can also be moved inside enable_training. Needs more investigation
+      training::GraphTransformerRegistry::GetInstance().RegisterExternalGraphTransformers();
```
Comment (souptc) [inline onnxruntime/core/session/environment.cc:271]: if i remember correctly, this api is added for Apollo usage, which we don't need it anymore. we can double confirm whether we can remove it, but i believe you don't need it for on device training.

**Review by souptc — APPROVED** [review]

File: onnxruntime/core/session/environment.cc (hunk 18 above)
Comment (askhade) [reply onnxruntime/core/session/environment.cc:271]: makes sense will cover this in a separate PR which I am working on right now... Will merge this PR.

**Review by pengwa — APPROVED** [review]

**Review by jchen351 — APPROVED** [review]
